### PR TITLE
SCMOD-5410: Worker Framework: Potential for task statuses not to be reported

### PR DESCRIPTION
**Pull request by gaana-vishwnath**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -103,6 +103,7 @@
         <module>worker-store-s3</module>
         <module>worker-store-mem</module>
         <module>worker-tracking-report</module>
+        <module>woker-testing</module>
```
Comment (andyreidz) [inline pom.xml]: Did you mean to add this?

File: worker-api/src/main/java/com/hpe/caf/api/worker/TaskInformation.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hpe.caf.api.worker;
+
+public interface TaskInformation {
+    String getInboundMessageId();
+    void incrementResponseCount(final boolean isFinalResponse);
```
Comment (andyreidz) [inline worker-api/src/main/java/com/hpe/caf/api/worker/TaskInformation.java]: We agreed to move incrementResponseCount to the RabbitTaskInformation

File: pom.xml (hunk 1 above)
Comment (gaana-vishwnath) [reply pom.xml]: sorry Andy, will be careful from now on. I missed this. Will remove it.

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hpe.caf.worker.queue.rabbit;
+
+import com.hpe.caf.api.worker.TaskInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RabbitTaskInformation implements TaskInformation {
```
Comment (andyreidz) [inline worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java:24]: Are you satisfied this is thread safe?

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            return responseCount;
+        }
+    }
+
+    public boolean isNegativeAckSent()
+    {
+        return negativeAckSent;
+    }
+
+    public void markNegativeAckAsSent()
```
Comment (andyreidz) [inline worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java]: We talked about a better name for this. isTaskRejected setTaskRejected

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitWorkerQueue.java [hunk 5]
```diff
@@ -17,14 +17,12 @@
...
-import com.hpe.caf.api.worker.ManagedWorkerQueue;
-import com.hpe.caf.api.worker.QueueException;
-import com.hpe.caf.api.worker.TaskCallback;
-import com.hpe.caf.api.worker.WorkerQueueMetricsReporter;
+import com.hpe.caf.api.worker.*;
 import com.hpe.caf.configs.RabbitConfiguration;
 import com.hpe.caf.util.rabbitmq.*;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
+import javafx.concurrent.Task;
```
Comment (andyreidz) [inline worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitWorkerQueue.java]: ?

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java [hunk 6]
```diff
@@ -77,32 +78,49 @@ public void handleAck(long sequenceNo, boolean multiple)
...
+                   return new ConsumerRejectEvent(Long.valueOf(t.getInboundMessageId()));
+               }
+               return null;
+           });
     }
 
-    private void handle(long sequenceNo, boolean multiple, Function<Long, Event<QueueConsumer>> eventSource)
+    private void handle(long sequenceNo, boolean multiple, Function<RabbitTaskInformation, Event<QueueConsumer>> eventSource)
     {
+        //TODO Andy Correct this logic
```
Comment (andyreidz) [inline worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java]: Is this TODO ready to be removed? Are there others?

File: worker-queue-rabbit/src/test/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListenerTest.java [hunk 7]
```diff
@@ -224,7 +256,94 @@ public void testDuplicateRegister()
...
+        WorkerConfirmListener conf = new WorkerConfirmListener(q);
+        RabbitTaskInformation rabbitTaskInfo_100 = new RabbitTaskInformation("100");
+        rabbitTaskInfo_100.incrementResponseCount(false);
+        conf.registerResponseSequence(5, rabbitTaskInfo_100);
+        rabbitTaskInfo_100.incrementResponseCount(false);
+        conf.registerResponseSequence(1,rabbitTaskInfo_100);
+        rabbitTaskInfo_100.incrementResponseCount(true);
+        conf.registerResponseSequence(2, rabbitTaskInfo_100);
+        conf.handleNack(4, true);
+        Event<QueueConsumer> e = q.poll(1000, TimeUnit.MILLISECONDS);
```
Comment (andyreidz) [inline worker-queue-rabbit/src/test/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListenerTest.java:297]: Is it worth verifying null is returned from q.poll after the relevant calls to registerResponseSequence. Same comment applies to other tests below.

**Review by andyreidz — CHANGES_REQUESTED** [review]
Before we merge to master we need to evaluate the impact on workers and document workers.

File: worker-queue-rabbit/src/test/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListenerTest.java (hunk 7 above)
Comment (gaana-vishwnath) [reply worker-queue-rabbit/src/test/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListenerTest.java:297]: yes Andy, in the next conf.handleAck(5, true); in line 303... you will not get any event as a negative ACK is already sent. Found it to be valid.

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java (hunk 6 above)
Comment (gaana-vishwnath) [reply worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java]: Nothing else is left, can be removed.

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitWorkerQueue.java (hunk 5 above)
Comment (gaana-vishwnath) [reply worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitWorkerQueue.java]: was updated

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java (hunk 4 above)
Comment (gaana-vishwnath) [reply worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java]: removed the setter and gave the name markNegativeAckAsSent(), thinking that to be a good enough name.

Comment (andyreidz) [reply worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/RabbitTaskInformation.java]: To me it's confusing, markNegativeAckAsSent when in reality a negativeAck was received.

File: worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java (hunk 6 above)
Comment (andyreidz) [reply worker-queue-rabbit/src/main/java/com/hpe/caf/worker/queue/rabbit/WorkerConfirmListener.java]: Please remove it

**Review by andyreidz — APPROVED** [review]

**Comment by andyreidz** [comment]
Also need the release notes updated.

**Pull request merged**